### PR TITLE
move modules to version 0.12

### PR DESCRIPTION
**Pull request by venturel**

### CODE REVIEW

File: instance/main.tf [hunk 1]
```diff
@@ -1,72 +1,104 @@
...
+    volume_type           = var.root_vl_type
+    volume_size           = var.root_vl_size
+    delete_on_termination = var.root_vl_delete
   }
 
-  ebs_block_device = ["${var.ebs_block_devices}"]
+  dynamic "ebs_block_device" {
+    for_each = [var.ebs_block_devices]
+    content {
+      # TF-UPGRADE-TODO: The automatic upgrade tool can't predict
```
Comment (Komorebi-E) [inline instance/main.tf]: Does this comment from tool need to stay?

File: puppet-userdata/outputs.tf [hunk 2]
```diff
@@ -1,3 +1,4 @@
 output "user_datas" {
-  value = ["${data.template_file.metadata_puppet.*.rendered}"]
+  value = [data.template_file.metadata_puppet.*.rendered]
```
Comment (Komorebi-E) [inline puppet-userdata/outputs.tf]: Are the explicit list brackets needed here?

File: bastion/main.tf [hunk 3]
```diff
@@ -1,64 +1,68 @@
+terraform {
```
Comment (iuriaranda) [inline bastion/main.tf]: can you move this to a separate `versions.tf` file?

File: bastion/main.tf [hunk 4]
```diff
@@ -1,64 +1,68 @@
...
+  termination_protection = var.termination_protection
+  root_vl_type           = var.root_vl_type
+  root_vl_size           = var.root_vl_size
+  root_vl_delete         = var.root_vl_delete
+  user_data              = var.user_data
   public_ip              = "true"
-  ebs_block_devices      = ["${var.ebs_block_devices}"]
-  ebs_enabled            = "${var.ebs_enabled}"
-  tags                   = "${var.tags}"
+  ebs_block_devices      = [var.ebs_block_devices]
```
Comment (iuriaranda) [inline bastion/main.tf]: Probably the list notation is not needed here

File: instance/iam.tf [hunk 5]
```diff
@@ -1,11 +1,11 @@
 resource "aws_iam_instance_profile" "profile" {
-  count = "${var.instance_count == "0" ? 0 : 1}"
+  count = var.instance_count == "0" ? 0 : 1
```
Comment (iuriaranda) [inline instance/iam.tf]: can you use the `number` type for this variable?

File: instance/main.tf (hunk 1 above)
Comment (iuriaranda) [reply instance/main.tf]: indeed the comment can go away. I think the actual implementation is ok

File: instance/outputs.tf [hunk 6]
```diff
@@ -1,47 +1,78 @@
 output "instance_ids" {
-  value = ["${concat(aws_instance.instance.*.id,aws_instance.instance_no_ebs.*.id)}"]
+  value = [concat(
```
Comment (iuriaranda) [inline instance/outputs.tf]: The list notation can be removed here as well

File: puppet-userdata/outputs.tf (hunk 2 above)
Comment (iuriaranda) [reply puppet-userdata/outputs.tf]: afaik no

File: bastion/main.tf [hunk 7]
```diff
@@ -2,63 +2,63 @@
...
-  sgs                    = "${concat(list("${aws_security_group.sg_bastion.id}"),"${var.sgs}")}"
+  name                   = var.name
+  subnets                = var.subnets
+  project                = var.project
+  environment            = var.environment
+  ami                    = var.ami
+  instance_type          = var.instance_type
+  key_name               = var.ssh_key_name
+  sgs                    = concat([aws_security_group.sg_bastion.id], var.sgs)
   instance_count         = "1"
```
Comment (iuriaranda) [inline bastion/main.tf]: this should be a number now

File: bastion/main.tf [hunk 8]
```diff
@@ -2,63 +2,63 @@
...
-  root_vl_type           = "${var.root_vl_type}"
-  root_vl_size           = "${var.root_vl_size}"
-  root_vl_delete         = "${var.root_vl_delete}"
-  user_data              = "${var.user_data}"
+  termination_protection = var.termination_protection
+  root_vl_type           = var.root_vl_type
+  root_vl_size           = var.root_vl_size
+  root_vl_delete         = var.root_vl_delete
+  user_data              = var.user_data
   public_ip              = "true"
```
Comment (iuriaranda) [inline bastion/main.tf]: this should be a bool, you might need to change the variable type as well in the instance module

File: instance/iam.tf (hunk 5 above)
Comment (iuriaranda) [reply instance/iam.tf]: you're still comparing it to a string 😛

File: README.md [hunk 9]
```diff
@@ -144,8 +144,8 @@ This module generates a script that bootstraps puppet on the server. It'll insta
 module "tools_userdata" {
   source              = "github.com/skyscrapers/terraform-instances//puppet-userdata?ref=1.0.1"
   amount_of_instances = "1"
```
Comment (iuriaranda) [inline README.md]: This should be a number as well

File: bastion/main.tf [hunk 10]
```diff
@@ -2,63 +2,63 @@
...
+  public_ip              = true
+  ebs_block_devices      = var.ebs_block_devices
+  ebs_enabled            = var.ebs_enabled
+  tags                   = var.tags
 }
 
 resource "aws_eip" "bastion_eip" {
   vpc      = true
-  instance = "${element(module.bastion_host.instance_ids, 0)}"
+  instance = element(module.bastion_host.instance_ids, 0)
```
Comment (duboisph) [inline bastion/main.tf]: `module.bastion_host.instance_ids[0]` should work too

File: instance/iam.tf [hunk 11]
```diff
@@ -1,11 +1,11 @@
 resource "aws_iam_instance_profile" "profile" {
-  count = "${var.instance_count == "0" ? 0 : 1}"
+  count = var.instance_count == 0 ? 0 : 1
```
Comment (duboisph) [inline instance/iam.tf]: Isn't instance count always > 0? What's the use case for having it set to 0?

File: instance/iam.tf [hunk 12]
```diff
@@ -1,11 +1,11 @@
...
   name  = "profile_${var.name}_${var.project}_${var.environment}"
-  role  = "${aws_iam_role.role.name}"
+  role  = aws_iam_role.role[0].name
 }
 
 resource "aws_iam_role" "role" {
-  count = "${var.instance_count == "0" ? 0 : 1}"
+  count = var.instance_count == 0 ? 0 : 1
```
Comment (duboisph) [inline instance/iam.tf]: Same question

File: instance/main.tf [hunk 13]
```diff
@@ -1,72 +1,96 @@
...
+    ]
   }
 
   credit_specification {
-    cpu_credits = "${var.cpu_credits}"
+    cpu_credits = var.cpu_credits
   }
 }
 
 resource "aws_instance" "instance_no_ebs" {
```
Comment (duboisph) [inline instance/main.tf]: THere shouldn't be a need anymore for separating resources depending on whether EBS is enabled. `The dynamic "ebs_block_device"` solves that

File: instance/outputs.tf [hunk 14]
```diff
@@ -1,47 +1,77 @@
 output "instance_ids" {
-  value = ["${concat(aws_instance.instance.*.id,aws_instance.instance_no_ebs.*.id)}"]
+  value = concat(
```
Comment (duboisph) [inline instance/outputs.tf]: See higher, with TF 0.12 we should simplify the module using dynamic blocks, so we don't need these workarounds anymore.

File: puppet-userdata/main.tf [hunk 15]
```diff
@@ -1,14 +1,14 @@
...
-    environment  = "${var.environment == "production" ? "" : "-${var.environment}"}"
-    project      = "${var.project == "" ? "" : "-${var.project}"}"
-    customer     = "${var.customer}"
-    function     = "${var.function}"
-    puppetmaster = "${var.puppetmaster}"
-    domain       = "${var.domain}"
+  vars = {
+    number       = count.index + 1
+    environment  = var.environment == "production" ? "" : "-${var.environment}"
+    project      = var.project == "" ? "" : "-${var.project}"
```
Comment (duboisph) [inline puppet-userdata/main.tf]: Maybe use `null` instead of empty string

File: instance/main.tf [hunk 16]
```diff
@@ -1,72 +1,96 @@
...
-  instance_type               = "${var.instance_type}"
-  key_name                    = "${var.key_name}"
-  iam_instance_profile        = "${aws_iam_instance_profile.profile.id}"
-  vpc_security_group_ids      = ["${var.sgs}"]
-  subnet_id                   = "${element(var.subnets, count.index)}"
-  disable_api_termination     = "${var.termination_protection}"
-  ebs_optimized               = "${module.is_ebs_optimised.is_ebs_optimised}"
-  associate_public_ip_address = "${var.public_ip}"
-  user_data                   = "${element(var.user_data, count.index)}"
+  count                       = var.ebs_enabled ? var.instance_count : 0
```
Comment (iuriaranda) [inline instance/main.tf]: I think we can now use `dynamic` blocks to simplify the module and use a single `aws_instance` resource for both use cases (with and without ebs).

File: instance/iam.tf (hunk 11 above)
Comment (venturel) [reply instance/iam.tf]: The idea behind it is to be able to disable a specific instance in a specific environment but e can remove this option (not sure if it should be part of this PR)

Comment (iuriaranda) [reply instance/iam.tf]: I agree that changing this should not be part of this PR

**Pull request merged**